src/amp.js

```diff
@@ -46,50 +49,58 @@ try {
 
   // Declare that this runtime will support a single root doc. Should happen
   // as early as possible.
-  /** @const {!./service/ampdoc-impl.AmpDocService} */
-  const ampdocService = installDocService(self, /* isSingleDoc */ true);
+  ampdocService = installDocService(self, /* isSingleDoc */ true);
+} catch (e) {
+  // In case of an error call this.
+  makeBodyVisible(self.document);
+  throw e;
+}
+chunk(self.document, function initial() {
   /** @const {!./service/ampdoc-impl.AmpDoc} */
   const ampdoc = ampdocService.getAmpDoc(self.document);
   /** @const {!./service/performance-impl.Performance} */
   const perf = installPerformanceService(self);
   perf.tick('is');
   installStyles(self.document, cssText, () => {
-    try {
+    chunk(self.document, function services() {
       // Core services.
       installRuntimeServices(self);
       installAmpdocServices(ampdoc);
       // We need the core services (viewer/resources) to start instrumenting
       perf.coreServicesAvailable();
       maybeTrackImpression(self);
-
+    });
+    chunk(self.document, function builtins() {
       // Builtins.
       installBuiltins(self);
+    });
+    chunk(self.document, function adoptWindow() {
+      try {
+        adopt(self);
+      } finally {
+        // Everything from here is nested because adopt itself
+        // may add additional chunks that should run first.
+        chunk(self.document, function stub() {
+          stubElements(self);
+        });
+        chunk(self.document, function final() {
+          installPullToRefreshBlocker(self);
+          installGlobalClickListenerForDoc(ampdoc);
 
-      // Final configuration and stubbing.
-      adopt(self);
-      stubElements(self);
-
-      installPullToRefreshBlocker(self);
-      installGlobalClickListenerForDoc(ampdoc);
-
-      maybeValidate(self);
-      makeBodyVisible(self.document, /* waitForServices */ true);
-      installCacheServiceWorker(self);
-    } catch (e) {
-      makeBodyVisible(self.document);
-      throw e;
-    } finally {
-      perf.tick('e_is');
-      // TODO(erwinm): move invocation of the `flush` method when we have the
-      // new ticks in place to batch the ticks properly.
-      perf.flush();
-    }
+          maybeValidate(self);
```

Comment:
Something like `maybeValidate` should definitely just be out of a critical path in its own microtask.